clippy_lints/src/field_reassign_with_default.rs

```diff
@@ -0,0 +1,208 @@
+
+        // start from the `let mut _ = _::default();` and look at all the following
+        // statements, see if they re-assign the fields of the binding
+        for (stmt_idx, binding_name, binding_type) in binding_statements_using_default {
+            // the last statement of a block cannot trigger the lint
+            if stmt_idx == block.stmts.len() - 1 {
+                break;
+            }
+
+            // find all "later statement"'s where the fields of the binding set as
+            // Default::default() get reassigned
+            let mut first_assign = None;
+            let mut assigned_fields = FxHashMap::default();
+            for consequtive_statement in &block.stmts[stmt_idx + 1..] {
+                // interrupt if the statement is a let binding (`Local`) that shadows the original
+                // binding
+                if stmt_shadows_binding(consequtive_statement, binding_name) {
+                    break;
+                }
+                // find out if and which field was set by this `consequtive_statement`
+                else if let Some((field_ident, assign_rhs)) =
+                    field_reassigned_by_stmt(consequtive_statement, binding_name)
+                {
+                    // extract and store the assigned value for help message
+                    let value_snippet = snippet(cx, assign_rhs.span, "..");
+
+                    // always re-insert set value, this way the latest value is stored for output snippet
+                    assigned_fields.insert(field_ident.name, value_snippet);
+
+                    // also set first instance of error for help message
+                    if first_assign.is_none() {
+                        first_assign = Some(consequtive_statement);
+                    }
+                }
+                // interrupt also if no field was assigned, since we only want to look at consequtive statements
+                else {
+                    break;
+                }
+            }
+
+            // if there are incorrectly assigned fields, do a span_lint_and_note to suggest
+            // construction using `Ty { fields, ..Default::default() }`
+            if !assigned_fields.is_empty() {
+                // take the original assignment as span
+                let stmt = &block.stmts[stmt_idx];
+
+                if let StmtKind::Local(preceding_local) = &stmt.kind {
+                    // if all fields of the struct are not assigned, add `.. Default::default()` to the suggestion.
+                    let ext_with_default = !fields_of_type(&binding_type)
+                        .iter()
+                        .all(|field| assigned_fields.contains_key(&field.name));
+
+                    let field_list = assigned_fields
+                        .into_iter()
+                        .map(|(field, value)| format!("{}: {}", field, value))
+                        .collect::<Vec<String>>()
+                        .join(", ");
```

Comment:
The order of the fields in the suggestion should match the order of the replaced statements. 

Consider this case:
```rust
    let mut y = A::default();
    y.j = order_matters() as i64;
    y.i = order_matters();
```
With a function like this:
```rust
fn order_matters() -> i32 {
    static mut VAL: i32 = 0;
    unsafe {
        VAL += 1;
        VAL
    }
}
```
The suggestion would change the values that `i` and `j` get.

Reply:
Order is now retained.